include/core/rtos/Thread.hpp

```diff
@@ -0,0 +1,312 @@
+
+#ifndef EVT_RTOS_THREAD_
+#define EVT_RTOS_THREAD_
+
+#include <core/rtos/Initializable.hpp>
+#include <cstdint>
+
+namespace core::rtos {
+
+/**
+ * Class that wraps a ThreadX Thread. Each thread represents a relatively independent task that will operate
+ * "concurrently" with other threads. Each thread has its own stack, which is allocated in a BytePool.\n\n
+ *
+ * This class extends Initializable, and like all other Initializable classes must be passed into the
+ * Threadx::startKernel() method as part of the initList parameter.
+ * @tparam T what type of data the thread's entry function will take. Should be at most 32 bits. (so generally should be
```

Comment:
The threadX method requires that thread entry functions take in a ulong (32 bit integer). I think they expect the user to cast that ulong to a (void*) then cast that void* into a pointer to whatever they actually want it to be, but using the template class lets the compiler handle all that and makes it very seamless to the end-coder.

However, I have no way to restrict the size of the template argument, which is a shame. So if someone passes in an entire struct instead of a pointer to a struct, it will cause some weird undefined behavior.
I can solve this by instead of expecting them to pass in a pointer as the template argument, have them pass in the object they want a pointer to
i.e.
currently, if they want to pass in a pointer to mystruct_t into the entry function:
EntryFunction(mystruct_t*) {...}
they say:
Thread<mystruct_t*>(...);
We could change it so if they want to use the same entry function:
EntryFunction(mystruct_t*) {...}
they say:
Thread<mystruct_t>(...); //note the lack of a pointer in the template argument

So we automatically assume they want a pointer to whatever template argument they give. This will solve the sizing issue, but, if they wanted to pass in something smaller than 32 bytes, like a uint32_t, they would be required to pass in a pointer to that uint32_t, which is just sorta weird.

Both methods have their downsides and upsides, and I'd be interested to hear what you think about either of them.
Alternatively, the most recent C++ version might have some way of restricting template arguments to a specific domain, but idk much about that.

Reply:
If you change the template to `template <typename T, typename = typename std::enable_if<sizeof(T) <= 4>::type>`, it will work. It would be better to use C++20 templates, but this does the job. If you're interested, you can read a bit here: https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error

I've just started learning about this stuff, and it's wild

Reply:
@mjh9585 What are your thoughts on upgrading to C++20 so we can get concepts?

Reply:
Upgrading would clean this up, but I think that is something that should be done in a separate pull request. This one is long enough without a version change. And with a version change we should verify nothing is broken because we were accidently relying on undefined behavior.

Reply:
Yeah, I think `enable_if` is good for now, and we can talk about C++20 later

Reply:
alright. Used enable_if and it seems to be working, i.e. it throws an error if you try to pass in something larger than 4 bytes as the argument. Thanks!